atomisticparsers/h5md/parser.py

```diff
@@ -0,0 +1,787 @@
+            for section in section_segments:
+                try:
+                    value = source.get(section)
+                    source = value[-1] if isinstance(value, list) else value
+                except Exception:
+                    return
+        value = source.attrs.get(attribute)
+        source = value[-1] if isinstance(value, list) else value
+        source = self.decode_hdf5_bytes(source, default) if source is not None else default
+
+        return source
+
+    def hdf5_getter(self, source, path, default=None):
+        '''
+        Extracts attribute from object based on path, and returns default if not defined.
+        '''
+        section_segments = path.split('.')
+        for section in section_segments:
+            try:
+                value = source.get(section)
+                unit = self.hdf5_attr_getter(source, 'unit', path=section)
+                unit_factor = self.hdf5_attr_getter(source, 'unit_factor', path=section, default=1.0)
+                source = value[-1] if isinstance(value, list) else value
+            except Exception:
+                return
+
+        if source is None:
+            source = default
+        elif type(source) == h5py.Dataset:
+            source = source[()]
+            source = self.apply_unit(source, unit, unit_factor)
+
+        return self.decode_hdf5_bytes(source, default)
+
+    def parse(self, quantity_key: str = None, **kwargs):
+        pass
+
+
+class H5MDParser(MDParser):
+    def __init__(self):
+        super().__init__()
+        self._data_parser = HDF5Parser()
+        self._n_frames = None  # TODO use trajectory_steps from MDParser?
+        self._n_atoms = None
+        self._frame_rate = None
+        self._atom_parameters = None
+        self._system_info = None
+        self._observable_info = None
+        self._parameter_info = None
+        self._h5md_groups = None
+        self._h5md_particle_group_all = None
+        self._h5md_positions_group_all = None
+        self._h5md_positions_value_all = None
+
+        self._nomad_to_particles_group_map = {
+            'positions': 'position',
+            'velocities': 'velocity',
+            'forces': 'force',
+            'labels': 'species_label',
+            'label': 'force_field_label',
+            'mass': 'mass',
+            'charge': 'charge',
+        }
+
+        self._nomad_to_box_group_map = {
+            'lattice_vectors': 'edges',
+            'periodic': 'boundary',
+            'dimension': 'dimension'
+        }
+
+    @property
+    def h5md_groups(self):
+        if self._h5md_groups is None:
+            if not self._data_parser.filehdf5:
+                return
+            groups = ['h5md', 'particles', 'observables', 'connectivity', 'parameters']
+            self._h5md_groups = {group: self._data_parser.hdf5_getter(self._data_parser.filehdf5, group) for group in groups}
+        return self._h5md_groups
+
+    @property
+    def h5md_particle_group_all(self):
```

Comment:
for defs l174, l183, l191, I suggest you do not make them as property, make them simply as private variable e.g. self._h5md_particle_group_all which you calculate at the beginning of def parse.

Reply:
Ah ok. got it.

Just a general question about this parse() function: I guess this is automatically triggered, but when/how exactly does this happen? When a class is instantiated?

Reply:
there are a lot of layers with the addition of plugin (plugin-MatchingParser-MatchingParserInterface-Parser)..the parser class is only instantiated once and parse is executed everytime you parse a mainfile so all intitializations should be here. The parser class is istantiated in l411 of nomad.parsing.parser.py and parse is called in l423 which is triggered in l1203 in nomad.processing.data.py